engine/execution/ingestion/finalized_and_executed_handler.go

```diff
@@ -0,0 +1,162 @@
+func NewFinalizedAndExecutedDispatcher(
+	latestFinalizedAndExecutedHeight uint64,
+	isBlockFinalized func(context.Context, *flow.Header) (bool, error),
+	isBlockExecuted func(context.Context, *flow.Header) (bool, error),
+	options ...FinalizedAndExecutedDispatcherOption,
+) FinalizedAndExecutedDispatcher {
+	d := &finalizedAndExecutedDispatcher{
+		log: zerolog.Nop(),
+
+		blockFinalizedAndExecutedConsumers: []OnBlockFinalizedAndExecutedConsumer{},
+		latestFinalizedAndExecutedHeight:   latestFinalizedAndExecutedHeight,
+
+		isBlockFinalized: isBlockFinalized,
+		isBlockExecuted:  isBlockExecuted,
+	}
+
+	for _, option := range options {
+		option(d)
+	}
+
+	return d
+}
+
+func (f *finalizedAndExecutedDispatcher) AddOnFinalizedAndExecutedConsumer(
+	consumer OnBlockFinalizedAndExecutedConsumer,
+) {
+	f.Lock()
+	defer f.Unlock()
+
+	f.blockFinalizedAndExecutedConsumers =
+		append(f.blockFinalizedAndExecutedConsumers, consumer)
+}
+
+func (f *finalizedAndExecutedDispatcher) OnBlockExecuted(
+	ctx context.Context,
+	h *flow.Header,
+) error {
+	isNew, err := f.checkOnBlockExecuted(ctx, h)
+	if !isNew || err != nil {
+		return err
+	}
+
+	f.dispatchFinalizedAndExecuted(h)
+
+	return nil
+}
+
+func (f *finalizedAndExecutedDispatcher) OnBlockFinalized(
+	ctx context.Context,
+	h *flow.Header,
+) error {
+	isNew, err := f.checkOnBlockFinalized(ctx, h)
+	if !isNew || err != nil {
+		return err
+	}
+
+	f.dispatchFinalizedAndExecuted(h)
+
+	return nil
+}
+
+// checkOnBlockExecuted checks if this is a new highest executed and finalized block
+// we know its executed because we are in the OnBlockExecuted callback
+func (f *finalizedAndExecutedDispatcher) checkOnBlockExecuted(
+	ctx context.Context,
+	h *flow.Header,
+) (bool, error) {
+	f.RLock()
+	defer f.RUnlock()
+
+	// we check latestFinalizedAndExecutedHeight+1 to further ensure that we are not
+	// skipping blocks
+	if h.Height != f.latestFinalizedAndExecutedHeight+1 {
+		return false, nil
+	}
+
+	if finalized, err := f.isBlockFinalized(ctx, h); !finalized || err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// checkOnBlockFinalized checks if this is a new highest executed and finalized block
+// we know its finalized because we are in the OnBlockFinalized callback
+func (f *finalizedAndExecutedDispatcher) checkOnBlockFinalized(
+	ctx context.Context,
+	h *flow.Header,
+) (bool, error) {
+	f.RLock()
+	defer f.RUnlock()
+
+	// we check latestFinalizedAndExecutedHeight+1 to further ensure that we are not
+	// skipping blocks
+	if h.Height != f.latestFinalizedAndExecutedHeight+1 {
+		return false, nil
+	}
+
+	if executed, err := f.isBlockExecuted(ctx, h); !executed || err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// dispatchFinalizedAndExecuted dispatches events to all registered consumers
+func (f *finalizedAndExecutedDispatcher) dispatchFinalizedAndExecuted(h *flow.Header) {
+	f.Lock()
+	defer f.Unlock()
```

Comment:
OnBlockFinalized and OnBlockExecuted happen concurrently, I'm not sure the fact that h was the latestFinalizedAndExecutedHeight is still true after the read lock is released and re-acquire here. 

And also not sure if we can guarantee `latestFinalizedAndExecutedHeight` is updated incrementally, and block header is delivered in a strong incrementing order to consumers